File: backend/schedule.py
from datetime import date, timedelta
from typing import Iterable, List, Optional

from .models import HolidayRange, PostponeRecord, LessonOverride, parse_date


def holiday_set(holiday_ranges: Iterable[HolidayRange]) -> set[date]:
    holidays = set()
    for holiday in holiday_ranges:
        start = parse_date(holiday.start_date)
        end = parse_date(holiday.end_date)
        if not start or not end:
            continue
        if end < start:
            start, end = end, start
        current = start
        while current <= end:
            holidays.add(current)
            current += timedelta(days=1)
    return holidays


def _next_weekday(after_date: date, weekday: int) -> date:
    days_ahead = (weekday - after_date.weekday()) % 7
    if days_ahead == 0:
        days_ahead = 7
    return after_date + timedelta(days=days_ahead)


def _find_next_available_weekly(
    after_date: date, weekday: int, scheduled: set[date], holidays: set[date]
) -> date:
    candidate = _next_weekday(after_date, weekday)
    while candidate in scheduled or candidate in holidays:
        candidate = _next_weekday(candidate, weekday)
    return candidate


def generate_weekly_schedule(
    start_date: date, weekday: int, lesson_total: int, holiday_ranges: Iterable[HolidayRange]
) -> List[date]:
    holidays = holiday_set(holiday_ranges)
    schedule = []
    current = start_date
    if current.weekday() != weekday:
        current = _next_weekday(current - timedelta(days=1), weekday)

    while len(schedule) < lesson_total:
        if current not in holidays:
            schedule.append(current)
        current = current + timedelta(days=7)

    return schedule


def apply_postpones(
    schedule: List[date],
    weekday: int,
    holiday_ranges: Iterable[HolidayRange],
    postpones: Iterable[PostponeRecord],
) -> List[date]:
    holidays = holiday_set(holiday_ranges)
    updated = list(schedule)
    scheduled = set(updated)

    for postpone in postpones:
        original = parse_date(postpone.original_date)
        if not original or original not in scheduled:
            continue
        scheduled.remove(original)
        updated.remove(original)

        target = parse_date(postpone.make_up_date)
        if not target or target in holidays or target in scheduled or target <= original:
            target = _find_next_available_weekly(original, weekday, scheduled, holidays)
        scheduled.add(target)
        updated.append(target)

    updated.sort()
    return updated


def apply_overrides(schedule: List[date], holiday_ranges: Iterable[HolidayRange], overrides: Iterable[LessonOverride]) -> List[date]:
    holidays = holiday_set(holiday_ranges)
    updated = list(schedule)
    for override in overrides:
        target = parse_date(override.date)
        if not target:
            continue
        if override.action == "remove":
            if target in updated:
                updated.remove(target)
        elif override.action == "add":
            if target not in updated and target not in holidays:
                updated.append(target)
    updated.sort()
    return updated


def calculate_progress(
    start_date_str: str,
    weekday: int,
    lesson_total: int,
    holiday_ranges: Iterable[HolidayRange],
    postpones: Iterable[PostponeRecord],
    overrides: Iterable[LessonOverride],
) -> dict:
    start_date = parse_date(start_date_str)
    if not start_date:
        return {
            "lessons_elapsed": 0,
            "lessons_remaining": lesson_total,
            "end_date": "",
            "next_lesson_date": "",
        }

    schedule = generate_weekly_schedule(start_date, weekday, lesson_total, holiday_ranges)
    schedule = apply_postpones(schedule, weekday, holiday_ranges, postpones)
    schedule = apply_overrides(schedule, holiday_ranges, overrides)

    today = date.today()
    elapsed = sum(1 for lesson in schedule if lesson <= today)
    remaining = max(lesson_total - elapsed, 0)
    end_date = schedule[-1].isoformat() if schedule else ""
    next_lesson = next((lesson for lesson in schedule if lesson >= today), None)
    next_lesson_date = next_lesson.isoformat() if next_lesson else ""

    return {
        "lessons_elapsed": elapsed,
        "lessons_remaining": remaining,
        "end_date": end_date,
        "next_lesson_date": next_lesson_date,
    }

File: src-tauri/src/schedule.rs
use chrono::{Datelike, Duration, NaiveDate};
use std::collections::HashSet;

use crate::models::{parse_date, HolidayRange, LessonOverride, PostponeRecord};

pub fn holiday_set(holiday_ranges: &[HolidayRange]) -> HashSet<NaiveDate> {
    let mut holidays = HashSet::new();
    for holiday in holiday_ranges {
        let start = match parse_date(&holiday.start_date) {
            Some(d) => d,
            None => continue,
        };
        let end = match parse_date(&holiday.end_date) {
            Some(d) => d,
            None => continue,
        };
        let (start, end) = if end < start { (end, start) } else { (start, end) };
        let mut current = start;
        while current <= end {
            holidays.insert(current);
            current += Duration::days(1);
        }
    }
    holidays
}

fn next_weekday(after_date: NaiveDate, weekday: u32) -> NaiveDate {
    let days_ahead = ((weekday as i64 - after_date.weekday().num_days_from_monday() as i64) % 7 + 7) % 7;
    let days_ahead = if days_ahead == 0 { 7 } else { days_ahead };
    after_date + Duration::days(days_ahead)
}

pub fn find_next_available_weekly(
    after_date: NaiveDate,
    weekday: u32,
    scheduled: &HashSet<NaiveDate>,
    holidays: &HashSet<NaiveDate>,
) -> NaiveDate {
    let mut candidate = next_weekday(after_date, weekday);
    while scheduled.contains(&candidate) || holidays.contains(&candidate) {
        candidate = next_weekday(candidate, weekday);
    }
    candidate
}

pub fn generate_weekly_schedule(
    start_date: NaiveDate,
    weekday: u32,
    lesson_total: i64,
    holiday_ranges: &[HolidayRange],
) -> Vec<NaiveDate> {
    let holidays = holiday_set(holiday_ranges);
    let mut schedule = Vec::new();
    let mut current = start_date;
    if current.weekday().num_days_from_monday() != weekday {
        current = next_weekday(current - Duration::days(1), weekday);
    }

    while (schedule.len() as i64) < lesson_total {
        if !holidays.contains(&current) {
            schedule.push(current);
        }
        current += Duration::days(7);
    }

    schedule
}

pub fn apply_postpones(
    schedule: &[NaiveDate],
    weekday: u32,
    holiday_ranges: &[HolidayRange],
    postpones: &[PostponeRecord],
) -> Vec<NaiveDate> {
    let holidays = holiday_set(holiday_ranges);
    let mut updated: Vec<NaiveDate> = schedule.to_vec();
    let mut scheduled: HashSet<NaiveDate> = updated.iter().cloned().collect();

    for postpone in postpones {
        let original = match parse_date(&postpone.original_date) {
            Some(d) => d,
            None => continue,
        };
        if !scheduled.contains(&original) {
            continue;
        }
        scheduled.remove(&original);
        updated.retain(|d| *d != original);

        let mut target = parse_date(&postpone.make_up_date);
        if let Some(t) = target {
            if holidays.contains(&t) || scheduled.contains(&t) || t <= original {
                target = None;
            }
        }
        let target = target.unwrap_or_else(|| {
            find_next_available_weekly(original, weekday, &scheduled, &holidays)
        });

        scheduled.insert(target);
        updated.push(target);
    }

    updated.sort();
    updated
}

pub fn apply_overrides(
    schedule: &[NaiveDate],
    holiday_ranges: &[HolidayRange],
    overrides: &[LessonOverride],
) -> Vec<NaiveDate> {
    let holidays = holiday_set(holiday_ranges);
    let mut updated: Vec<NaiveDate> = schedule.to_vec();
    for ov in overrides {
        let target = match parse_date(&ov.date) {
            Some(d) => d,
            None => continue,
        };
        if ov.action == "remove" {
            updated.retain(|d| *d != target);
        } else if ov.action == "add" {
            if !updated.contains(&target) && !holidays.contains(&target) {
                updated.push(target);
            }
        }
    }
    updated.sort();
    updated
}

pub fn calculate_progress(
    start_date_str: &str,
    weekday: u32,
    lesson_total: i64,
    holiday_ranges: &[HolidayRange],
    postpones: &[PostponeRecord],
    overrides: &[LessonOverride],
) -> serde_json::Value {
    let start_date = match parse_date(start_date_str) {
        Some(d) => d,
        None => {
            return serde_json::json!({
                "lessons_elapsed": 0,
                "lessons_remaining": lesson_total,
                "end_date": "",
                "next_lesson_date": "",
            });
        }
    };

    let schedule = generate_weekly_schedule(start_date, weekday, lesson_total, holiday_ranges);
    let schedule = apply_postpones(&schedule, weekday, holiday_ranges, postpones);
    let schedule = apply_overrides(&schedule, holiday_ranges, overrides);

    let today = chrono::Local::now().date_naive();
    let elapsed = schedule.iter().filter(|&&d| d <= today).count() as i64;
    let remaining = (lesson_total - elapsed).max(0);
    let end_date = schedule.last().map(|d| d.format("%Y-%m-%d").to_string()).unwrap_or_default();
    let next_lesson = schedule.iter().find(|&&d| d >= today);
    let next_lesson_date = next_lesson.map(|d| d.format("%Y-%m-%d").to_string()).unwrap_or_default();

    serde_json::json!({
        "lessons_elapsed": elapsed,
        "lessons_remaining": remaining,
        "end_date": end_date,
        "next_lesson_date": next_lesson_date,
    })
}
